### PR TITLE
WT-9765 Fix handling a prepared commit after a prepared rollback with eviction failure

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/txn/txn.c [hunk 1]
```diff
@@ -1231,8 +1231,14 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
       (upd->type != WT_UPDATE_TOMBSTONE ||
         (upd->next != NULL && upd->durable_ts == upd->next->durable_ts &&
           upd->txnid == upd->next->txnid && upd->start_ts == upd->next->start_ts));
-    first_committed_upd_in_hs =
-      first_committed_upd != NULL && F_ISSET(first_committed_upd, WT_UPDATE_HS);
+    /*
+     * If the first committed update after the prepared transaction has already been marked to be
+     * deleted from the history store, no need to do anything because there is another prepared
+     * update that is rolled back before the current prepared update.
+     */
```
Comment (kommiharibabu) [inline src/txn/txn.c:1238]: Can you add a simple scenario explaining it would be much easier to understand?

File: test/suite/test_prepare25.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            if self.delete:
+                self.session.begin_transaction()
+                cursor.set_key(i)
+                cursor.remove()
+                self.session.commit_transaction('commit_timestamp=' + self.timestamp_str(ts + 20))
+
+            # Do a prepared update
+            self.session.begin_transaction()
+            cursor[i] = value_b
+            self.session.prepare_transaction('prepare_timestamp=' + self.timestamp_str(ts + 30))
```
Comment (kommiharibabu) [inline test/suite/test_prepare25.py:85]: The cursor is not closed here. It pins the page that avoids eviction. Close the cursor.

Comment (quchenhao) [reply test/suite/test_prepare25.py:85]: done

File: src/txn/txn.c (hunk 1 above)
Comment (quchenhao) [reply src/txn/txn.c:1244]: done

**Review by kommiharibabu — APPROVED** [review]
LGTM

File: src/txn/txn.c [hunk 3]
```diff
@@ -1231,8 +1231,20 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
       (upd->type != WT_UPDATE_TOMBSTONE ||
         (upd->next != NULL && upd->durable_ts == upd->next->durable_ts &&
           upd->txnid == upd->next->txnid && upd->start_ts == upd->next->start_ts));
-    first_committed_upd_in_hs =
-      first_committed_upd != NULL && F_ISSET(first_committed_upd, WT_UPDATE_HS);
+    /*
+     * If the first committed update after the prepared update has already been marked to be
```
Comment (Siddhartha8899) [inline src/txn/txn.c:1235]: Shall we make it clear that this comment provides the reasoning for the `WT_UPDATE_TO_DELETE_FROM_HS ` flag check? 

Also, shouldn't the comment say:
` If the first committed update **before** the prepared update... `

**Review by Siddhartha8899 — CHANGES_REQUESTED** [review]

File: src/txn/txn.c [hunk 4]
```diff
@@ -1231,8 +1231,20 @@ __txn_resolve_prepared_op(WT_SESSION_IMPL *session, WT_TXN_OP *op, bool commit,
...
+     * deleted from the history store, we are in the case that there was a previous prepared update
```
Comment (Siddhartha8899) [inline src/txn/txn.c:1236]: same here:
`we are in the case that there was a **next** prepared update . . .`

**Review by Siddhartha8899 — CHANGES_REQUESTED** [review]

File: test/suite/test_prepare25.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    delete = [
+        ('delete', dict(delete=True)),
+        ('non-delete', dict(delete=False)),
+    ]
+
+    scenarios = make_scenarios(format_values, delete)
+
+    def test_prepare25(self):
+        uri = "table:test_prepare25"
```
Comment (Siddhartha8899) [inline test/suite/test_prepare25.py:51]: Just curious if this test fails without your changes.

File: src/txn/txn.c (hunk 3 above)
Comment (quchenhao) [reply src/txn/txn.c:1235]: It should be first committed update older than the prepared upate

File: src/txn/txn.c (hunk 4 above)
Comment (quchenhao) [reply src/txn/txn.c:1236]: I don't understand what you mean by next?

File: src/txn/txn.c (hunk 3 above)
Comment (Siddhartha8899) [reply src/txn/txn.c]: Yup, that is better. Thanks for the update.

File: src/txn/txn.c (hunk 4 above)
Comment (Siddhartha8899) [reply src/txn/txn.c]: I will resolve this as the updated comment makes sense.

**Review by Siddhartha8899 — APPROVED** [review]

**Comment by agorrod** [comment]
@quchenhao I forget what this is up to - is it OK to merge?

**Comment by quchenhao** [comment]
@agorrod Yes, I'll merge this after WT-9705 is merged.

**Comment by quchenhao** [comment]
evergreen merge

**Pull request merged**